s3fs/core.py

```diff
@@ -1433,38 +1435,63 @@ def flush(self, force=False, retries=10):
             if force:
                 self.forced = True
 
-            self.buffer.seek(0)
-            part = len(self.parts) + 1
-            i = 0
-
             try:
                 self.mpu = self.mpu or self._call_s3(
                     self.s3.s3.create_multipart_upload,
                     Bucket=self.bucket, Key=self.key, ACL=self.acl)
-            except (ClientError, ParamValidationError) as e:
-                raise_from(IOError('Initiating write failed: %s' %
-                                   self.path), e)
+            except (ClientError, ParamValidationError) as exc:
+                raise_from(IOError('Initiating write failed: %s' % self.path), exc)
 
-            while True:
-                try:
-                    out = self._call_s3(
-                        self.s3.s3.upload_part,
-                        Bucket=self.bucket,
-                        PartNumber=part, UploadId=self.mpu['UploadId'],
-                        Body=self.buffer.read(), Key=self.key)
-                    break
-                except S3_RETRYABLE_ERRORS:
-                    if i < retries:
-                        logger.debug('Exception %e on S3 write, retrying',
-                                     exc_info=True)
-                        i += 1
-                        continue
+            self.buffer.seek(0)
+
+            # chunk buffer to ensure blocksize respected by multiparts;
+            # look ahead to ensure we don't attempt to write too small or too large a part
+            (data0, data1) = (None, self.buffer.read(self.blocksize))
+            while data1:
+                (data0, data1) = (data1, self.buffer.read(self.blocksize))
+
+                data1_size = len(data1)
+                if 0 < data1_size < self.blocksize:
+                    # last part would be smaller than specified
+                    remainder = data0 + data1
+                    remainder_size = self.blocksize + data1_size
+
+                    if remainder_size <= self.part_max:
+                        # we can combine the last two parts without issue
+                        (data0, data1) = (remainder, None)
+                    else:
+                        # combination of last two blocks is too large for a single part
```

Comment:
This is the extreme edge case?

Reply:
Indeed. I believe you'd have to set the `blocksize` to more than 2.5 GiB in order to hit it. But, so long as that's configurable, and so long as we're deciding what size parts to send here, it seems wise to handle.